optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractLeftHandSide.java

```diff
@@ -67,4 +69,91 @@ protected AbstractLeftHandSide(DroolsVariableFactory variableFactory) {
+     * @param prerequisitePattern never null, pattern required to construct the variable
+     * @param boundVarA never null, {@link TriTuple#a}
+     * @param boundVarB never null, {@link TriTuple#b}
+     * @param boundVarC never null, {@link TriTuple#c}
+     * @param <A> generic type of the first bound variable
+     * @param <B> generic type of the second bound variable
+     * @param <C> generic type of the third bound variable
+     * @return never null
+     */
+    protected static <A, B, C> DirectPatternVariable<TriTuple<A, B, C>> decompose(
+            Variable<TriTuple<A, B, C>> primaryVariable, ViewItem<?> prerequisitePattern, Variable<A> boundVarA,
+            Variable<B> boundVarB, Variable<C> boundVarC) {
+        return new DirectPatternVariable<>(primaryVariable, singletonList(prerequisitePattern))
+                .bind(boundVarA, tuple -> tuple.a)
+                .bind(boundVarB, tuple -> tuple.b)
+                .bind(boundVarC, tuple -> tuple.c);
+    }
+
+    /**
+     * Create a {@link DirectPatternVariable} on {@link QuadTuple} with pre-made bindings for its components variables.
+     *
+     * @param primaryVariable never null
+     * @param prerequisitePattern never null, pattern required to construct the variable
+     * @param boundVarA never null, {@link QuadTuple#a}
+     * @param boundVarB never null, {@link QuadTuple#b}
+     * @param boundVarC never null, {@link QuadTuple#c}
+     * @param boundVarD never null, {@link QuadTuple#d}
+     * @param <A> generic type of the first bound variable
+     * @param <B> generic type of the second bound variable
+     * @param <C> generic type of the third bound variable
+     * @param <D> generic type of the fourth bound variable
+     * @return never null
+     */
+    protected static <A, B, C, D> DirectPatternVariable<QuadTuple<A, B, C, D>> decompose(
+            Variable<QuadTuple<A, B, C, D>> primaryVariable, ViewItem<?> prerequisitePattern, Variable<A> boundVarA,
+            Variable<B> boundVarB, Variable<C> boundVarC, Variable<D> boundVarD) {
+        return new DirectPatternVariable<>(primaryVariable, singletonList(prerequisitePattern))
+                .bind(boundVarA, tuple -> tuple.a)
+                .bind(boundVarB, tuple -> tuple.b)
+                .bind(boundVarC, tuple -> tuple.c)
+                .bind(boundVarD, tuple -> tuple.d);
+    }
+
+    protected static ViewItem<?> buildAccumulate(ViewItem<?> innerAccumulatePattern,
+            AccumulateFunction... accFunctions) {
+        if (accFunctions.length == 0) {
+            throw new IllegalStateException("Impossible state: no accumulate functions provided.");
+        } else if (accFunctions.length == 1) {
+            return DSL.accumulate(innerAccumulatePattern, accFunctions[0]);
+        } else {
+            return DSL.accumulate(innerAccumulatePattern, accFunctions[0],
+                    Arrays.stream(accFunctions)
+                            .skip(1)
```

Comment:
Would you please explain why we skip the first element?

Reply:
Because we've already used it on the previous line.

Reply:
Yes, I've just overlooked it.